R/add_cols.R

```diff
@@ -118,60 +118,97 @@ NULL
 }
 
 #' @name .add_date
-.add_deaths <- function(.data,
-                        onset_to_death,
-                        hosp_death_risk,
-                        non_hosp_death_risk) {
+.add_outcome <- function(.data,
+                         onset_to_death,
+                         onset_to_recovery,
+                         hosp_death_risk,
+                         non_hosp_death_risk,
+                         config) {
   infected_idx <- .data$infected == "infected"
   num_infected <- sum(infected_idx)
-  .data$deaths <- NA_real_
-  .data$deaths[infected_idx] <- .data$time[infected_idx] +
-    onset_to_death(num_infected)
+  .data$outcome <- "contact"
+  .data$outcome_time <- NA_real_
+  .data$outcome[infected_idx] <- "recovered"
+  .data$outcome_time[infected_idx] <- .data$time[infected_idx] +
+    onset_to_recovery(num_infected)
+  hosp_idx <- !is.na(.data$hospitalisation) & infected_idx
+  non_hosp_idx <- is.na(.data$hospitalisation) & infected_idx
 
-  apply_death_risk <- function(.data, risk, hosp = TRUE) {
+  # define functions used only by .add_outcome
+  time_varying_risk <- function(.data, died_idx, config) {
+    # check time-varying function
+    .check_func_req_args(config$time_varying_death_risk)
+
+    # sample which deaths to keep with time-varying risk
+    prob_death <- config$time_varying_death_risk(.data$time[died_idx])
+    max_prob_death <- stats::optimise(
+      f = config$time_varying_death_risk,
+      interval = c(0, 100),
+      maximum = TRUE
+    )$objective
+    norm_prob_death <- prob_death / max_prob_death
+    # set all values > 1 and < 0 to 1 and 0 due to imprecision of optimise
+    # TODO: ideally this step would be removed as no values are 0 > x < 1
+    norm_prob_death[norm_prob_death > 1] <- 1
+    norm_prob_death[norm_prob_death < 0] <- 0
+    sampled_died <- stats::rbinom(
+      n = length(died_idx),
+      size = 1,
+      prob = norm_prob_death
+    )
+    # subset deaths given sampling (logical conversion for vec subsetting)
+    died_idx <- died_idx[as.logical(sampled_died)]
+
+    # return sample of deaths
+    died_idx
+  }
+
+  apply_death_risk <- function(.data, risk, idx, config) {
     if (!is_na(risk)) {
```

Comment:
Thanks for pointing this out. I was unaware that {rlang} exported this function. In fact, the implementation I've added is equivalent to `rlang::is_lgl_na()`, as my `is_na()` function fails with typed `NA`s (e.g. `NA_character_`).

{rlang} is already a package dependency of {simulist} so using the `is_lgl_na()` wouldn't increase the number of dependencies.

The lifecycle is questioning as it might be moved to {vctrs}. This would be annoying if I started using the function as I'd then have to either take on {vctrs} or reimplement my own function. 

I'll address this in a separate PR, but will likely move to using the {rlang} function so thanks for pointing it out.

Reply:
I assumed that the 'questioning' status was why this was replicated - and I think that remains an okay reason to have this to avoid importing {vctrs}.